Consistent General Protection Fault; zfs_rangelock_enter() -> avl_nearest() -> avl_walk()

### System information
Type | Version/Name
 --- | ---
Distribution Name	| Proxmox
Distribution Version	|  Proxmox 6.2 (Debian Buster)
Linux Kernel	| 5.4.44-2-pve
Architecture	| x86_64
ZFS Version	| 0.8.4-pve1
SPL Version	| 0.8.4-pve1

### Describe the problem you're observing
I'm experiencing a consistent general protection fault after upgrade from 0.7.x (Proxmox 5.4.x) to 0.8.4 (Proxmox 6.2). This fault looks similar to issue #7873, but not exactly the same. The result of this fault is a complete hang of most processes running within the Linux Container that uses the associated ZFS volume, along with various other unresponsive host processes. I cannot regain use of the container until the entire Proxmox host is (hard) rebooted. The normal reboot process fails to proceed due to the hung LXC.

### Describe how to reproduce the problem
ZFS is my root (rpool), two JBOD drives mirrored, with several Linux Containers using their own ZFS volumes. 

Bug can be encountered by using the system normally; I can make it maybe two days before the issue forces me to reboot the machine.

I believe Proxmox has a scheduled scrub that - while running - increases the chance of occurrence. However, daily use also seems to cause the same issue. This has been consistently triggered by the `Plex Media Server` process for me; I have not seen any other processes noted in the errors, but I do not see how this can be a Plex issue (all fingers point to a ZFS issue).

I use ECC memory and have run memtest86+ recently to verify that memory is good.
Drives are healthy, pool is as well.

### Include any warning/errors/backtraces from the system logs
```log
[136422.505658] general protection fault: 0000 [#1] SMP PTI
[136422.505866] CPU: 15 PID: 25510 Comm: Plex Media Serv Tainted: P          IO      5.4.44-2-pve #1
[136422.506115] Hardware name: HP ProLiant ML350 G6, BIOS D22 05/05/2011
[136422.506303] RIP: 0010:avl_walk+0x33/0x70 [zavl]
[136422.506437] Code: 10 b9 01 00 00 00 29 d1 4c 01 c6 48 89 e5 48 85 f6 74 48 48 63 d2 48 89 f7 48 8b 04 d6 48 85 c0 74 20 48 63 c9 eb 03 48 89 d0 <48> 8b 14 c8 48 85 d2 75 f4 48 89 c2 48 89 d0 5d 4c 29 c0 c3 39 f1
[136422.507020] RSP: 0018:ffffacf29a163c80 EFLAGS: 00010282
[136422.507171] RAX: e089443875c085d0 RBX: ffffa0aac4d9c150 RCX: 0000000000000000
[136422.507375] RDX: 0000000000000001 RSI: ffffffffc01c1478 RDI: ffffffffc01c1478
[136422.507579] RBP: ffffacf29a163c80 R08: 0000000000000008 R09: ffffa0a8c3406f40
[136422.507784] R10: ffffa0ab33be8000 R11: 0000008000000000 R12: ffffa0ab33be8000
[136422.508031] R13: ffffa0aac4d9c178 R14: 0000000000000000 R15: 0000000000000000
[136422.508236] FS:  00007f6f78ff9700(0000) GS:ffffa0adc39c0000(0000) knlGS:0000000000000000
[136422.508465] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[136422.508630] CR2: 00007f6fbc294278 CR3: 0000000469bfc000 CR4: 00000000000006e0
[136422.508834] Call Trace:
[136422.508910]  avl_nearest+0x2a/0x30 [zavl]
[136422.509140]  zfs_rangelock_enter+0x405/0x580 [zfs]
[136422.509286]  ? spl_kmem_zalloc+0xe9/0x140 [spl]
[136422.509419]  ? spl_kmem_zalloc+0xe9/0x140 [spl]
[136422.509587]  zfs_get_data+0x157/0x340 [zfs]
[136422.509746]  zil_commit_impl+0x9ad/0xd90 [zfs]
[136422.509913]  zil_commit+0x3d/0x60 [zfs]
[136422.510062]  zfs_fsync+0x77/0xe0 [zfs]
[136422.510252]  zpl_fsync+0x68/0xa0 [zfs]
[136422.510364]  vfs_fsync_range+0x48/0x80
[136422.510474]  ? __fget_light+0x59/0x70
[136422.510582]  do_fsync+0x3d/0x70
[136422.510674]  __x64_sys_fsync+0x14/0x20
[136422.510786]  do_syscall_64+0x57/0x190
[136422.510895]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[136422.511041] RIP: 0033:0x7f6fffef4b17
[136422.511146] Code: 00 00 0f 05 48 3d 00 f0 ff ff 77 3f f3 c3 0f 1f 44 00 00 53 89 fb 48 83 ec 10 e8 04 f5 ff ff 89 df 89 c2 b8 4a 00 00 00 0f 05 <48> 3d 00 f0 ff ff 77 2b 89 d7 89 44 24 0c e8 46 f5 ff ff 8b 44 24
[136422.511700] RSP: 002b:00007f6f78ff79c0 EFLAGS: 00000293 ORIG_RAX: 000000000000004a
[136422.511916] RAX: ffffffffffffffda RBX: 000000000000000c RCX: 00007f6fffef4b17
[136422.512121] RDX: 0000000000000000 RSI: 0000000000000002 RDI: 000000000000000c
[136422.512373] RBP: 00000000038164b8 R08: 0000000000000000 R09: 00007f6fbc447770
[136422.512583] R10: 00007f6fbc0c4460 R11: 0000000000000293 R12: 0000000000000000
[136422.512792] R13: 0000000003825eb8 R14: 0000000000000002 R15: 0000000000000000
[136422.512999] Modules linked in: veth ebtable_filter ebtables ip_set ip6table_raw iptable_raw ip6table_filter ip6_tables mptctl mptbase binfmt_misc iptable_filter bpfilter softdog cpuid nfnetlink_log nfnetlink radeon ttm drm_kms_helper drm i2c_algo_bit fb_sys_fops syscopyarea sysfillrect sysimgblt intel_powerclamp usblp kvm_intel ipmi_ssif input_leds kvm hpilo i7core_edac pcspkr irqbypass serio_raw ipmi_si intel_cstate ipmi_devintf ipmi_msghandler mac_hid vhost_net vhost tap ib_iser rdma_cm iw_cm ib_cm ib_core iscsi_tcp libiscsi_tcp libiscsi scsi_transport_iscsi coretemp parport_pc ppdev lp parport sunrpc ip_tables x_tables autofs4 zfs(PO) zunicode(PO) zlua(PO) zavl(PO) icp(PO) zcommon(PO) znvpair(PO) spl(O) btrfs xor zstd_compress raid6_pq libcrc32c hid_generic usbkbd usbmouse usbhid hid gpio_ich psmouse pata_acpi uhci_hcd ehci_pci lpc_ich tg3 ehci_hcd hpsa scsi_transport_sas
[136422.515330] ---[ end trace b7f8cd7b0f091d95 ]---
[136422.515472] RIP: 0010:avl_walk+0x33/0x70 [zavl]
[136422.515649] Code: 10 b9 01 00 00 00 29 d1 4c 01 c6 48 89 e5 48 85 f6 74 48 48 63 d2 48 89 f7 48 8b 04 d6 48 85 c0 74 20 48 63 c9 eb 03 48 89 d0 <48> 8b 14 c8 48 85 d2 75 f4 48 89 c2 48 89 d0 5d 4c 29 c0 c3 39 f1
[136422.516197] RSP: 0018:ffffacf29a163c80 EFLAGS: 00010282
[136422.516352] RAX: e089443875c085d0 RBX: ffffa0aac4d9c150 RCX: 0000000000000000
[136422.516577] RDX: 0000000000000001 RSI: ffffffffc01c1478 RDI: ffffffffc01c1478
[136422.516826] RBP: ffffacf29a163c80 R08: 0000000000000008 R09: ffffa0a8c3406f40
[136422.517030] R10: ffffa0ab33be8000 R11: 0000008000000000 R12: ffffa0ab33be8000
[136422.517235] R13: ffffa0aac4d9c178 R14: 0000000000000000 R15: 0000000000000000
[136422.517439] FS:  00007f6f78ff9700(0000) GS:ffffa0adc39c0000(0000) knlGS:0000000000000000
[136422.517669] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[136422.517875] CR2: 00007f6fbc294278 CR3: 0000000469bfc000 CR4: 00000000000006e0
```

@tuxoko 

I successfully built the .deb packages, but I'm having trouble at the dkms module build stage. I'm seeing if I can get things to work by rebuilding with `CFLAGS=-Wno-error`, unless there's something better I should be doing?

```
/var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.c: In function ‘zfs_get_data’:
/var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.c:1068:3: error: implicit declaration of function ‘zfs_zrele_async’; did you mean ‘zfs_iput_async’? [-Werror=implicit-function-declaration]
   zfs_zrele_async(zp);
   ^~~~~~~~~~~~~~~
   zfs_iput_async
/var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.c:1073:40: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 3 has type ‘uint64_t’ {aka ‘long long unsigned int’} [-Wformat=]
   printk("%s: gen mismatch, expected=%lu got=%lu\n", __func__, gen, zp_gen);
                                      ~~^                       ~~~
                                      %llu
/var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.c:1073:48: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 4 has type ‘uint64_t’ {aka ‘long long unsigned int’} [-Wformat=]
   printk("%s: gen mismatch, expected=%lu got=%lu\n", __func__, gen, zp_gen);
                                              ~~^                    ~~~~~~
                                              %llu
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zio_checksum.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zio_compress.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zio_crypt.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zio_inject.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zle.o
cc1: some warnings being treated as errors
make[7]: *** [/usr/src/linux-headers-4.19.0-13-common/scripts/Makefile.build:308: /var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.o] Error 1
make[7]: *** Waiting for unfinished jobs....
make[6]: *** [/usr/src/linux-headers-4.19.0-13-common/scripts/Makefile.build:549: /var/lib/dkms/zfs/0.8.6/build/module/zfs] Error 2
make[5]: *** [/usr/src/linux-headers-4.19.0-13-common/Makefile:1565: _module_/var/lib/dkms/zfs/0.8.6/build/module] Error 2
make[4]: *** [Makefile:146: sub-make] Error 2
make[3]: *** [Makefile:8: all] Error 2
make[3]: Leaving directory '/usr/src/linux-headers-4.19.0-13-amd64'
make[2]: *** [Makefile:30: modules] Error 2
make[2]: Leaving directory '/var/lib/dkms/zfs/0.8.6/build/module'
make[1]: *** [Makefile:843: all-recursive] Error 1
make[1]: Leaving directory '/var/lib/dkms/zfs/0.8.6/build'
make: *** [Makefile:712: all] Error 2
```
Cheers

@tuxoko 

I'm trying to install your patch on `zfs-dkms` but I'm unable to build the `dkms` module properly, because of `-Werror=implicit-function-declaration`. Where would be the proper place for me to add `-Wall -Wno-error` to `CFLAGS` (or whatever flags are needed to override) as part of the `dkms` module build process? I tried doing this both with an environment variable while calling `dkms install`, and setting in the Makefile `/var/lib/dkms/zfs/0.8.6/build/Makefile`, but I think this needs to be declared somewhere in `/usr/src/zfs-0.8.6`?


```make
/var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.c: In function ‘zfs_get_data’:
/var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.c:1068:3: error: implicit declaration of function ‘zfs_zrele_async’; did you mean ‘zfs_iput_async’? [-Werror=implicit-function-declaration]
   zfs_zrele_async(zp);
   ^~~~~~~~~~~~~~~
   zfs_iput_async
/var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.c:1073:40: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 3 has type ‘uint64_t’ {aka ‘long long unsigned int’} [-Wformat=]
   printk("%s: gen mismatch, expected=%lu got=%lu\n", __func__, gen, zp_gen);
                                      ~~^                       ~~~
                                      %llu
/var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.c:1073:48: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 4 has type ‘uint64_t’ {aka ‘long long unsigned int’} [-Wformat=]
   printk("%s: gen mismatch, expected=%lu got=%lu\n", __func__, gen, zp_gen);
                                              ~~^                    ~~~~~~
                                              %llu
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zio_checksum.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zio_compress.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zio_crypt.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zio_inject.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zle.o
cc1: some warnings being treated as errors
make[7]: *** [/usr/src/linux-headers-4.19.0-13-common/scripts/Makefile.build:308: /var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.o] Error 1
make[7]: *** Waiting for unfinished jobs....
make[6]: *** [/usr/src/linux-headers-4.19.0-13-common/scripts/Makefile.build:549: /var/lib/dkms/zfs/0.8.6/build/module/zfs] Error 2
make[5]: *** [/usr/src/linux-headers-4.19.0-13-common/Makefile:1565: _module_/var/lib/dkms/zfs/0.8.6/build/module] Error 2
make[4]: *** [Makefile:146: sub-make] Error 2
make[3]: *** [Makefile:8: all] Error 2
make[3]: Leaving directory '/usr/src/linux-headers-4.19.0-13-amd64'
make[2]: *** [Makefile:30: modules] Error 2
make[2]: Leaving directory '/var/lib/dkms/zfs/0.8.6/build/module'
make[1]: *** [Makefile:843: all-recursive] Error 1
make[1]: Leaving directory '/var/lib/dkms/zfs/0.8.6/build'
make: *** [Makefile:712: all] Error 2
```

Cheers

@tuxoko Oh weird, I never really thought about that, that is interesting that gcc can do that.

After correcting that, I'm still getting some dkms build errors, but different ones this time:

```make
/var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.c: In function ‘zfs_get_data’:
/var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.c:1068:18: error: passing argument 1 of ‘zfs_iput_async’ from incompatible pointer type [-Werror=incompatible-pointer-types]
   zfs_iput_async(zp);
                  ^~
/var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.c:992:30: note: expected ‘struct inode *’ but argument is of type ‘znode_t *’ {aka ‘struct znode *’}
 zfs_iput_async(struct inode *ip)
                ~~~~~~~~~~~~~~^~
/var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.c:1075:18: error: passing argument 1 of ‘zfs_iput_async’ from incompatible pointer type [-Werror=incompatible-pointer-types]
   zfs_iput_async(zp);
                  ^~
/var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.c:992:30: note: expected ‘struct inode *’ but argument is of type ‘znode_t *’ {aka ‘struct znode *’}
 zfs_iput_async(struct inode *ip)
                ~~~~~~~~~~~~~~^~
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zil.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zio.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zio_checksum.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zio_compress.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zio_crypt.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zio_inject.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zle.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zpl_ctldir.o
  CC [M]  /var/lib/dkms/zfs/0.8.6/build/module/zfs/zpl_export.o
cc1: some warnings being treated as errors
make[7]: *** [/usr/src/linux-headers-4.19.0-13-common/scripts/Makefile.build:308: /var/lib/dkms/zfs/0.8.6/build/module/zfs/zfs_vnops.o] Error 1
make[7]: *** Waiting for unfinished jobs....
make[6]: *** [/usr/src/linux-headers-4.19.0-13-common/scripts/Makefile.build:549: /var/lib/dkms/zfs/0.8.6/build/module/zfs] Error 2
make[5]: *** [/usr/src/linux-headers-4.19.0-13-common/Makefile:1565: _module_/var/lib/dkms/zfs/0.8.6/build/module] Error 2
make[4]: *** [Makefile:146: sub-make] Error 2
make[3]: *** [Makefile:8: all] Error 2
make[3]: Leaving directory '/usr/src/linux-headers-4.19.0-13-amd64'
make[2]: *** [Makefile:30: modules] Error 2
make[2]: Leaving directory '/var/lib/dkms/zfs/0.8.6/build/module'
make[1]: *** [Makefile:843: all-recursive] Error 1
make[1]: Leaving directory '/var/lib/dkms/zfs/0.8.6/build'
make: *** [Makefile:712: all] Error 2
```

I'm guessing there's a few more tweaks that need to be done?

Also is this commit related to this issue at all?
https://github.com/openzfs/zfs/commit/2921ad6cba54522a33fc95f90237c6e5ead80de7

--- a/cmd/ztest/ztest.c
+++ b/cmd/ztest/ztest.c
@@ -2287,8 +2287,8 @@ ztest_get_done(zgd_t *zgd, int error)
 }
 
 static int
-ztest_get_data(void *arg, lr_write_t *lr, char *buf, struct lwb *lwb,
-    zio_t *zio)
+ztest_get_data(void *arg, uint64_t arg2, lr_write_t *lr, char *buf,
+    struct lwb *lwb, zio_t *zio)
 {
 	ztest_ds_t *zd = arg;
 	objset_t *os = zd->zd_os;

--- a/include/sys/zil.h
+++ b/include/sys/zil.h
@@ -399,6 +399,7 @@ typedef struct itx {
 	void		*itx_callback_data; /* User data for the callback */
 	size_t		itx_size;	/* allocated itx structure size */
 	uint64_t	itx_oid;	/* object id */
+	uint64_t	itx_gen;	/* gen number for zfs_get_data */
 	lr_t		itx_lr;		/* common part of log record */
 	/* followed by type-specific part of lr_xx_t and its immediate data */
 } itx_t;
@@ -467,7 +468,7 @@ typedef int zil_parse_blk_func_t(zilog_t *zilog, const blkptr_t *bp, void *arg,
 typedef int zil_parse_lr_func_t(zilog_t *zilog, const lr_t *lr, void *arg,
     uint64_t txg);
 typedef int zil_replay_func_t(void *arg1, void *arg2, boolean_t byteswap);
-typedef int zil_get_data_t(void *arg, lr_write_t *lr, char *dbuf,
+typedef int zil_get_data_t(void *arg, uint64_t arg2, lr_write_t *lr, char *dbuf,
     struct lwb *lwb, zio_t *zio);
 
 extern int zil_parse(zilog_t *zilog, zil_parse_blk_func_t *parse_blk_func,

--- a/include/sys/zvol_impl.h
+++ b/include/sys/zvol_impl.h
@@ -85,8 +85,8 @@ void zvol_log_truncate(zvol_state_t *zv, dmu_tx_t *tx, uint64_t off,
     uint64_t len, boolean_t sync);
 void zvol_log_write(zvol_state_t *zv, dmu_tx_t *tx, uint64_t offset,
     uint64_t size, int sync);
-int zvol_get_data(void *arg, lr_write_t *lr, char *buf, struct lwb *lwb,
-    zio_t *zio);
+int zvol_get_data(void *arg, uint64_t arg2, lr_write_t *lr, char *buf,
+    struct lwb *lwb, zio_t *zio);
 int zvol_init_impl(void);
 void zvol_fini_impl(void);
 void zvol_wait_close(zvol_state_t *zv);

--- a/module/zfs/zfs_log.c
+++ b/module/zfs/zfs_log.c
@@ -540,6 +540,7 @@ zfs_log_write(zilog_t *zilog, dmu_tx_t *tx, int txtype,
 	uint32_t blocksize = zp->z_blksz;
 	itx_wr_state_t write_state;
 	uintptr_t fsync_cnt;
+	uint64_t gen = 0;
 
 	if (zil_replaying(zilog, tx) || zp->z_unlinked ||
 	    zfs_xattr_owner_unlinked(zp)) {
@@ -561,6 +562,9 @@ zfs_log_write(zilog_t *zilog, dmu_tx_t *tx, int txtype,
 	if ((fsync_cnt = (uintptr_t)tsd_get(zfs_fsyncer_key)) != 0) {
 		(void) tsd_set(zfs_fsyncer_key, (void *)(fsync_cnt - 1));
 	}
+
+	(void) sa_lookup(zp->z_sa_hdl, SA_ZPL_GEN(ZTOZSB(zp)), &gen,
+	    sizeof (gen));
 
 	while (resid) {
 		itx_t *itx;
@@ -609,6 +613,7 @@ zfs_log_write(zilog_t *zilog, dmu_tx_t *tx, int txtype,
 		BP_ZERO(&lr->lr_blkptr);
 
 		itx->itx_private = ZTOZSB(zp);
+		itx->itx_gen = gen;
 
 		if (!(ioflag & (O_SYNC | O_DSYNC)) && (zp->z_sync_cnt == 0) &&
 		    (fsync_cnt == 0))

--- a/module/zfs/zfs_vnops.c
+++ b/module/zfs/zfs_vnops.c
@@ -738,7 +738,8 @@ static void zfs_get_done(zgd_t *zgd, int error);
  * Get data to generate a TX_WRITE intent log record.
  */
 int
-zfs_get_data(void *arg, lr_write_t *lr, char *buf, struct lwb *lwb, zio_t *zio)
+zfs_get_data(void *arg, uint64_t gen, lr_write_t *lr, char *buf,
+    struct lwb *lwb, zio_t *zio)
 {
 	zfsvfs_t *zfsvfs = arg;
 	objset_t *os = zfsvfs->z_os;
@@ -749,6 +750,7 @@ zfs_get_data(void *arg, lr_write_t *lr, char *buf, struct lwb *lwb, zio_t *zio)
 	dmu_buf_t *db;
 	zgd_t *zgd;
 	int error = 0;
+	uint64_t zp_gen;
 
 	ASSERT3P(lwb, !=, NULL);
 	ASSERT3P(zio, !=, NULL);
@@ -764,6 +766,16 @@ zfs_get_data(void *arg, lr_write_t *lr, char *buf, struct lwb *lwb, zio_t *zio)
 		 * Release the vnode asynchronously as we currently have the
 		 * txg stopped from syncing.
 		 */
+		zfs_zrele_async(zp);
+		return (SET_ERROR(ENOENT));
+	}
+	/* check if generation number matches */
+	if (sa_lookup(zp->z_sa_hdl, SA_ZPL_GEN(zfsvfs), &zp_gen,
+	    sizeof (zp_gen)) != 0) {
+		zfs_zrele_async(zp);
+		return (SET_ERROR(EIO));
+	}
+	if (zp_gen != gen) {
 		zfs_zrele_async(zp);
 		return (SET_ERROR(ENOENT));
 	}

--- a/module/zfs/zil.c
+++ b/module/zfs/zil.c
@@ -1744,7 +1744,8 @@ cont:
 			 * completed after "lwb_write_zio" completed.
 			 */
 			error = zilog->zl_get_data(itx->itx_private,
-			    lrwb, dbuf, lwb, lwb->lwb_write_zio);
+			    itx->itx_gen, lrwb, dbuf, lwb,
+			    lwb->lwb_write_zio);
 
 			if (error == EIO) {
 				txg_wait_synced(zilog->zl_dmu_pool, txg);

--- a/module/zfs/zvol.c
+++ b/module/zfs/zvol.c
@@ -660,7 +660,8 @@ zvol_get_done(zgd_t *zgd, int error)
  * Get data to generate a TX_WRITE intent log record.
  */
 int
-zvol_get_data(void *arg, lr_write_t *lr, char *buf, struct lwb *lwb, zio_t *zio)
+zvol_get_data(void *arg, uint64_t arg2, lr_write_t *lr, char *buf,
+    struct lwb *lwb, zio_t *zio)
 {
 	zvol_state_t *zv = arg;
 	uint64_t offset = lr->lr_offset;

